resharper/src/resharper-unity/CSharp/Daemon/Stages/PerformanceCriticalCodeAnalysis/PerformanceCriticalCodeAnalysisStage.cs

```diff
@@ -80,28 +82,111 @@ private void AnalyzeFile(ICSharpFile file, IHighlightingConsumer consumer)
+            HighlightCostlyReachableInvocation(consumer, context);
+        }
+
+        private static void HighlightCostlyReachableInvocation(IHighlightingConsumer consumer, HotMethodAnalyzerContext context)
+        {
+            foreach (var kvp in context.InvocationsInMethods)
+            {
+                var method = kvp.Key;
+                if (!context.IsDeclaredElementCostlyReachable(method))
+                    continue;
+
+                var invocationElements = kvp.Value;
+                foreach (var invocationElement in invocationElements)
+                {
+                    var invokedMethod = invocationElement.Key;
+                    if (!context.IsDeclaredElementCostlyReachable(invokedMethod))
+                        continue;
+
+                    var references = invocationElement.Value;
+                    foreach (var reference in references)
+                    {
+                        consumer.AddHighlighting(new PerformanceCriticalCodeInvocationReachableHighlighting(reference));
+                    }
+                }
+            }
+        }
+
+        private void PropagateCostlyReachableMark(HotMethodAnalyzerContext context)
+        {
+            var callGraph = context.InvertedCallGraph;
+            var nodes = new HashSet<IDeclaredElement>();
+
+            foreach (var (from, toVertices) in callGraph)
+            {
+                if (context.IsDeclaredElementCostlyReachable(from))
+                    nodes.Add(from);
+                foreach (var to in toVertices)
+                {
+                    if (context.IsDeclaredElementCostlyReachable(to))
+                        nodes.Add(to);
+                }
+            }
+
+            var visited = new HashSet<IDeclaredElement>();
+            foreach (var node in nodes)
+            {
+                PropagateCostlyMark(node, callGraph, context, visited);
+            }
+        }
+
+        private static HotMethodAnalyzerContext GetHotMethodAnalyzerContext(IHighlightingConsumer consumer, List<IMethodDeclaration> hotRootMethods,
+            IPsiSourceFile sourceFile)
+        {
+            // sharing context for each hot root.
+            var context = new HotMethodAnalyzerContext();
+
+            foreach (var methodDeclaration in hotRootMethods)
+            {
+                var declaredElement = methodDeclaration.DeclaredElement.NotNull("declaredElement != null");
+                context.CurrentDeclaredElement = declaredElement;
+
+                var visitor = new HotMethodAnalyzer(sourceFile, consumer, ourMaxAnalysisDepth);
+                methodDeclaration.ProcessDescendants(visitor, context);
+            }
+
+            return context;
+        }
+
+        private static List<IMethodDeclaration> FindHotRootMethods([NotNull]ICSharpFile file,[NotNull] IPsiSourceFile sourceFile)
+        {
+            var result = new List<IMethodDeclaration>();
```

Comment:
Often it's good to use `LocaList<T>` instead of `List<T>`, it doesn't allocate without need. Once you populated local list you can then invoke `.ReadOnlyList()` on it to get `IReadOnlyList<T>`.